torch/optim/optimizer.py

```diff
@@ -233,17 +233,73 @@ class Optimizer:
+                             "parameter groups. They should be equal.")
+        param_lens = (len(g['params']) for g in groups)
+        saved_lens = (len(g['params']) for g in saved_groups)
+        if any(p_len != s_len for p_len, s_len in zip(param_lens, saved_lens)):
+            raise ValueError("loaded state dict contains a parameter group "
+                             "that doesn't match the size of optimizer's group")
+
+        return dict(zip(chain.from_iterable(g['params'] for g in saved_groups),
+                        chain.from_iterable(g['params'] for g in groups)))
+
+
+    @staticmethod
+    def _cast_state_to_match_params_hook(optimizer, state_dict) -> None:
+        """The following hook is automatically registered to load_state_dict as a pre-hook. This processing
+        used to live within load_state_dict, but, since the introduction of state_dict hooks, we've moved
+        this to be its own hook to allow users flexibility to trigger pre-hook before OR after casting
+        state to match the params' dtype and device.
+        """
+
+        # Validate the state_dict
+        groups = optimizer.param_groups
+        saved_groups = state_dict['param_groups']
+        id_map = Optimizer._verify_param_groups_align(groups, saved_groups)
+
+        def _cast(param, value, param_id=None, param_groups=None, key=None):
+            r"""Make a deep copy of value, casting all tensors to device of param."""
+            if isinstance(value, torch.Tensor):
+                return Optimizer._process_value_according_to_param_policy(param, value, param_id, param_groups, key)
+            elif isinstance(value, dict):
+                return {k: _cast(param, v, param_id=param_id, param_groups=param_groups, key=k)
+                        for k, v in value.items()}
+            elif isinstance(value, Iterable):
+                return type(value)(_cast(param, v, param_id=param_id, param_groups=param_groups)
+                                   for v in value)  # type: ignore[call-arg]
+            else:
+                return value
+
+        # Cast state tensors to appropriate types in the state_dict
+        for k, v in state_dict['state'].items():
+            if k in id_map:
+                param = id_map[k]
+                state_dict['state'][k] = _cast(param, v, param_id=k, param_groups=saved_groups)
+
+
     def __init__(self, params: params_t, defaults: Dict[str, Any]) -> None:
         torch._C._log_api_usage_once("python.optimizer")
         self.defaults = defaults
         self._optimizer_step_pre_hooks = OrderedDict()
         self._optimizer_step_post_hooks = OrderedDict()
         self._optimizer_state_dict_pre_hooks = OrderedDict()
         self._optimizer_state_dict_post_hooks = OrderedDict()
-        self._optimizer_load_state_dict_pre_hooks = OrderedDict()
+        # _optimizer_load_state_dict_pre_hooks and _load_state_dict_cast_handle
```

Comment:
Comments explaining why they are not here is good. Please make all 3 consistent though